src/main/java/org/sonarlint/intellij/core/ServerIssueUpdater.java

```diff
@@ -0,0 +1,146 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonarlint.intellij.core;
+
+import com.intellij.openapi.components.AbstractProjectComponent;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import org.sonarlint.intellij.config.project.SonarLintProjectSettings;
+import org.sonarlint.intellij.issue.IssuePointer;
+import org.sonarlint.intellij.issue.IssueStore;
+import org.sonarlint.intellij.issue.ServerIssuePointer;
+import org.sonarlint.intellij.ui.SonarLintConsole;
+import org.sonarlint.intellij.util.SonarLintUtils;
+import org.sonarsource.sonarlint.core.client.api.connected.ConnectedSonarLintEngine;
+import org.sonarsource.sonarlint.core.client.api.connected.ServerIssue;
+import org.sonarsource.sonarlint.core.client.api.exceptions.SonarLintWrappedException;
+
+public class ServerIssueUpdater extends AbstractProjectComponent {
+
+  private static final Logger LOGGER = Logger.getInstance(ServerIssueUpdater.class);
+
+  private static final int THREADS_NUM = 5;
+  private static final int QUEUE_LIMIT = 20;
+
+  private ExecutorService executorService;
+
+  private final IssueStore store;
+
+  public ServerIssueUpdater(Project project, IssueStore store) {
+    super(project);
+    this.store = store;
+  }
+
+  public void fetchAndMatchServerIssues(VirtualFile virtualFile) {
+    SonarLintProjectSettings projectSettings = SonarLintUtils.get(myProject, SonarLintProjectSettings.class);
+    if (!projectSettings.isBindingEnabled()) {
+      // not in connected mode
+      return;
+    }
+
+    String serverId = projectSettings.getServerId();
+    String moduleKey = projectSettings.getProjectKey();
+    if (serverId == null || moduleKey == null) {
+      // not bound to SQ project
```

Comment:
And perhaps this part should be moved into SonarLintServerManager. It could have a method #getConnectedEngine(Project), and we extract the code doing the consistency checks in #getFacadeForAnalysis into a separate method